sdk-typed-search-attributes.md

```diff
@@ -0,0 +1,497 @@
+  * Search attributes getting on list/describe - as `Map<java.lang.String,​ ?>` on list but no high-level describe form
+    (from latest javadoc, though didn't check repo where we may have new describe form)
+* TS
+  * Search attributes getting in workflow - on info as `Record<string, scalar[]>` (i.e. object)
+  * Search attributes setting in workflow - via `upsertSearchAttributes` accepting `Record<string, scalar[]>`
+  * Search attributes setting on start - via `Record<string, scalar[]>` on start options
+  * Search attributes getting on list/describe - as `Record<string, scalar[]>`
+* Python
+  * Search attributes getting in workflow - on info as `Mapping[str, List[scalar]]`
+  * Search attributes setting in workflow - via `upsert_search_attributes` accepting `Mapping[str, List[scalar]]`
+  * Search attributes setting on start - via `Mapping[str, List[scalar]]` on start options
+  * Search attributes getting on list/describe - as `Mapping[str, List[scalar]]`
+* .NET
+  * Still in alpha, can decide whatever we want
+
+### Server
+
+* ElasticSearch
+  * Multi-value search attribute for non-keyword-list - accepted, but discouraged, and returned as given
+    * TODO(cretz): Question outstanding to server team about the benefits of multi-value non-keyword-list attributes
+  * Single-value search attribute for non-keyword-list - accepted and returned as given
+  * Multi-value search attribute for keyword-list - accepted and returned as given
+  * Single-value search attribute for keyword-list - fails
+* SQL
+  * Multi-value search attribute for non-keyword-list - only allowed if a single value, always returned single value
+  * Single-value search attribute for non-keyword-list - accepted and returned as given
+  * Multi-value search attribute for keyword-list - accepted and returned as given
+  * Single-value search attribute for keyword-list - fails
+
+### Problems with current behavior
+
+* Go and Java rely on user to resolve ambiguity between single and multi value search attributes (Go is worse because it has no helpers)
+* TypeScript and Python always disambiguate making all search attributes a list (and Java too in single-value cases) even if list not allowed anymore
+* Setting invalid search attributes in workflow never returns a failure that can be handled by workflow code
+* Users have no way to clarify the difference between keyword and keyword-list
+
+## Goals
+
+Immediate:
+
+* As much as is supported by the language, users need to be able to get and set individually typed search attributes
+* Maintain backwards compatibility but deprecate any approach that doesn't let you identify the search attribute type
+  alongside the use of its value
+
+Non-immediate:
+
+* Provide ways to define predefined search attributes that are typed
+* Also provide ways to define predefined memo values that are typed
+
+Non-goals:
+
+* Better validation of actual search attribute type from server
+* Better way for server to tell SDK the upsert failed
```

Comment:
My concern with ignoring this for now is this design would mean we'd have to make another different API if we ever want users to be able to upsert-and-confirm.

Reply:
Yes, we will have to decide whether returning errors on search attribute mutation is worth it in the long run.

If we do decide that it is, I need to change a lot of things here. Basically we'd be back to "upsert search attributes" which is fine, just need to get more input that that's the ideal direction (and we'd probably call it "apply search attributes" or something since we can't mess w/ that existing function).

Reply:
Right, the potential future would be to tell a user it failed as a normal failure they can react to in code. But if we don't think we'd ever add that, then we can not concern ourselves with it.